### PR TITLE
[FL-2230] SubGhz: protocol API refactoring

### PR DESCRIPTION
**Pull request by Skorpionm**
# What's new

- [FL-2230] SubGhz: protocol API refactoring
- Flipper format: seek_to_end, key_exist
- Flipper Format: insert_or_update

# Verification 

- Сompile and download
- Check the full operation of SubGhz, receiving, transmitting, saving, loading, recognizing and other functions of SubGhz operation.

# Checklist (do not modify)

- [X] PR has description of feature/bug or link to Confluence/Jira task
- [X] Description contains actions to verify feature/bugfix
- [X] I've built this code, uploaded it to the device and verified feature/bugfix

[FL-2230]: https://flipperzero.atlassian.net/browse/FL-2230?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: applications/subghz/views/subghz_frequency_analyzer.c [hunk 1]
```diff
@@ -6,30 +6,30 @@
 #include <furi_hal.h>
 #include <input/input.h>
 #include <notification/notification_messages.h>
-#include <lib/subghz/protocols/subghz_protocol_princeton.h>
+//#include <lib/subghz/protocols/subghz_protocol_princeton.h>
```
Comment (DrZlo13) [inline applications/subghz/views/subghz_frequency_analyzer.c]: Do not comment unused code.

File: applications/subghz/views/subghz_frequency_analyzer.c [hunk 2]
```diff
@@ -77,7 +77,7 @@ void subghz_frequency_analyzer_draw(Canvas* canvas, SubghzFrequencyAnalyzerModel
 
 bool subghz_frequency_analyzer_input(InputEvent* event, void* context) {
     furi_assert(context);
-    //SubghzFrequencyAnalyzer* instance = context;
+    //SubGhzFrequencyAnalyzer* instance = context;
```
Comment (DrZlo13) [inline applications/subghz/views/subghz_frequency_analyzer.c]: Same

File: applications/subghz/views/subghz_read_raw.c [hunk 3]
```diff
@@ -6,15 +6,15 @@
 #include <furi_hal.h>
 #include <input/input.h>
 #include <gui/elements.h>
-#include <lib/subghz/protocols/subghz_protocol_princeton.h>
+//#include <lib/subghz/protocols/subghz_protocol_princeton.h>
```
Comment (DrZlo13) [inline applications/subghz/views/subghz_read_raw.c]: Same

File: applications/subghz/views/subghz_read_raw.c [hunk 4]
```diff
@@ -65,7 +65,7 @@ void subghz_read_raw_add_data_rssi(SubghzReadRAW* instance, float rssi) {
     //if(u_rssi > 34) u_rssi = 34;
```
Comment (DrZlo13) [inline applications/subghz/views/subghz_read_raw.c]: And again...

File: lib/subghz/blocks/const.c [hunk 5]
```diff
@@ -0,0 +1 @@
+#include "const.h"
```
Comment (DrZlo13) [inline lib/subghz/blocks/const.c:1]: Do we really need an empty file?

File: lib/subghz/blocks/encoder.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <stddef.h>
+
+#include <lib/toolbox/level_duration.h>
+
+//typedef LevelDuration (*SubGhzProtocolCommonEncoderCallback)(void* context);
```
Comment (DrZlo13) [inline lib/subghz/blocks/encoder.h]: And same again, cleanup please.

File: lib/subghz/blocks/encoder.c [hunk 7]
```diff
@@ -0,0 +1,3 @@
+#include "encoder.h"
+
+#define TAG "SubGhzBlockEncoder"
```
Comment (DrZlo13) [inline lib/subghz/blocks/encoder.c:3]: Empty?

File: lib/subghz/blocks/math.h [hunk 8]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <stddef.h>
+
+#define bit_read(value, bit) (((value) >> (bit)) & 0x01)
```
Comment (DrZlo13) [inline lib/subghz/blocks/math.h:7]: I propose to move all bitwise operations to common_defines.h

File: applications/subghz/views/transmitter.c [hunk 9]
```diff
@@ -112,42 +112,43 @@ bool subghz_transmitter_input(InputEvent* event, void* context) {
...
     furi_assert(context);
-    // SubghzTransmitter* subghz_transmitter = context;
+    // SubGhzViewTransmitter* subghz_transmitter = context;
 }
 
-void subghz_transmitter_exit(void* context) {
+void subghz_view_transmitter_exit(void* context) {
     furi_assert(context);
-    // SubghzTransmitter* subghz_transmitter = context;
+    // SubGhzViewTransmitter* subghz_transmitter = context;
```
Comment (DrZlo13) [inline applications/subghz/views/transmitter.c]: Again

File: applications/subghz/views/transmitter.c [hunk 10]
```diff
@@ -112,42 +112,43 @@ bool subghz_transmitter_input(InputEvent* event, void* context) {
...
     }
 
     return true;
 }
 
-void subghz_transmitter_enter(void* context) {
+void subghz_view_transmitter_enter(void* context) {
     furi_assert(context);
-    // SubghzTransmitter* subghz_transmitter = context;
+    // SubGhzViewTransmitter* subghz_transmitter = context;
```
Comment (DrZlo13) [inline applications/subghz/views/transmitter.c]: And again

**Review by DrZlo13 — CHANGES_REQUESTED** [review]

File: lib/subghz/blocks/math.h (hunk 8 above)
Comment (DrZlo13) [reply lib/subghz/blocks/math.h:7]: I think now isn't time for that.

File: lib/subghz/blocks/const.c (hunk 5 above)
Comment (Skorpionm) [reply lib/subghz/blocks/const.c:1]: This file will be needed later.

File: lib/subghz/blocks/encoder.c (hunk 7 above)
Comment (Skorpionm) [reply lib/subghz/blocks/encoder.c:3]: This file will be needed later.

**Review by DrZlo13 — APPROVED** [review]

**Pull request merged**